server/jetstream_cluster.go

```diff
@@ -3223,6 +3227,176 @@ func (js *jetStream) applyStreamEntries(mset *stream, ce *CommittedEntry, isReco
 			}
 		}
 	}
+
+	// If we're still actively processing a batch, must store the entry in-memory
+	// to come back to it later once we find the commit.
+	if batchActiveId != _EMPTY_ {
+		mset.batchMu.Lock()
+		if mset.batchEntries == nil {
+			mset.batchEntries = []*CommittedEntry{ce}
+		} else {
+			mset.batchEntries = append(mset.batchEntries, ce)
+		}
+		maxApplied := mset.batchMaxApplied
+		mset.batchMu.Unlock()
+		return maxApplied, nil
+	}
+	return 0, nil
+}
+
+// clearBatchStateLocked clears in-memory apply-batch-related state.
+// mset.batchMu lock should be held.
+func (mset *stream) clearBatchStateLocked() {
```

Comment:
Maybe intro no Locked version that acquire the proper lock and release and call these functions.